lib/Dialect/Pulse/Utils/Utils.cpp

```diff
@@ -35,4 +45,94 @@ Waveform_CreateOp getWaveformOp(PlayOp pulsePlayOp,
   return wfrOp;
 }
 
+Waveform_CreateOp
+getWaveformOp(PlayOp pulsePlayOp,
+              std::deque<mlir::pulse::CallSequenceOp> &callSequenceOpStack) {
+  auto wfrIndex = 0;
+  mlir::Value wfrOp = pulsePlayOp.getWfr();
+
+  for (auto it = callSequenceOpStack.rbegin(); it != callSequenceOpStack.rend();
+       ++it) {
+    if (wfrOp.isa<BlockArgument>()) {
+      wfrIndex = wfrOp.dyn_cast<BlockArgument>().getArgNumber();
+      wfrOp = it->getOperand(wfrIndex);
+    } else
+      break;
+  }
+
+  auto waveformOp =
+      dyn_cast<mlir::pulse::Waveform_CreateOp>(wfrOp.getDefiningOp());
```

Comment:
Since these functions are in the `mlir::pulse` name space you should be able to drop it here and make the code a bit more readable. 
```suggestion
      dyn_cast<Waveform_CreateOp>(wfrOp.getDefiningOp());
```